### PR TITLE
Adds Nimbus events [SDK-61, SDK-65, SDK-126]

### PR DESCRIPTION
**Pull request by travis79**
This adds and emits the [Glean events for the Nimbus telemetry](https://mana.mozilla.org/wiki/pages/viewpage.action?spaceKey=FJT&title=Nimbus+Engineering#NimbusEngineering-ExperimentTelemetry). 

**Note** this requires data review before landing!!

### CODE REVIEW

**Comment by travis79** [comment]
# Request for data collection review form

**All questions are mandatory. You must receive review from a data steward peer on your responses to these questions before shipping new data collection.**

1) What questions will you answer with this data?

- We will use these events to monitor experiment enrollment and exposure.

2) Why does Mozilla need to answer these questions?  Are there benefits for users? Do we need this information to address product or business requirements? 

- We need to know this information to evaluate experiment enrollment and for analysis of the results of the experiment.

3) What alternative methods did you consider to answer these questions? Why were they not sufficient?

- N/A

4) Can current instrumentation answer these questions?

- No

5) List all proposed measurements and indicate the category of data collection for each measurement, using the [Firefox data collection categories](https://wiki.mozilla.org/Firefox/Data_Collection) found on the Mozilla wiki.   

**Note that the data steward reviewing your request will characterize your data collection based on the highest (and most sensitive) category.**

<table>
  <tr>
    <td>Measurement Description</td>
    <td>Data Collection Category</td>
    <td>Tracking Bug #</td>
  </tr>
  <tr>
    <td>Experiment enrollment event</td>
    <td>Category 1 (technical)</td>
    <td>[Jira issue](https://jira.mozilla.com/browse/SDK-61)</td>
  </tr>
  <tr>
    <td>Experiment unenrollment event</td>
    <td>Category 1 (technical)</td>
    <td>[Jira issue](https://jira.mozilla.com/browse/SDK-126)</td>
  </tr>
  <tr>
    <td>Experiment exposure event</td>
    <td>Category 1 (technical)</td>
    <td>[Jira issue](https://jira.mozilla.com/browse/SDK-65)</td>
  </tr>
  <tr>
    <td>Experiment disqualification event</td>
    <td>Category 1 (technical)</td>
    <td>[Jira issue](https://jira.mozilla.com/browse/SDK-61)</td>
  </tr>
</table>

6) Please provide a link to the documentation for this data collection which describes the ultimate data set in a public, complete, and accurate way.

- This collection is documented in its [metrics.md document](https://github.com/mozilla-mobile/android-components/tree/master/components/service/nimbus/docs/metrics.md). Since this hasn't landed yet you can preview the generated metrics.md [here](https://gist.github.com/8046b7e95746a8296175526369f577d0)

7) How long will this data be collected?  Choose one of the following:

- I want to permanently monitor this data. tlong@mozilla.com

8) What populations will you measure?

- All consumers of the Nimbus AC component

9) If this data collection is default on, what is the opt-out mechanism for users?

- Fenix provides an opt-out mechanism through its UI.

10) Please provide a general description of how you will analyze this data.

- Jetstream, Cirrus, STMO, etc.

11) Where do you intend to share the results of your analysis?

- Mozilla internal

12) Is there a third-party tool (i.e. not Telemetry) that you are proposing to use for this data collection? If so:

- N/A

**Comment by travis79** [comment]
Flagging @tdsmith for datareview

**Comment by tdsmith** [comment]
We should get @litkaemily to approve because this contains a new identifier, enrollmentId. The new identifier is scoped to the lifetime of an experiment. It is similar to the experiment telemetry we collect on the desktop platform. @travis79, can you comment on the behavior of the identifier with respect to client_id?

The rest lgtm pending Emily's response.

1) Is there or will there be **documentation** that describes the schema for the ultimate data set in a public, complete, and accurate way?

Yes, in components/service/nimbus/docs/metrics.md and the Glean probe dictionary.

2) Is there a control mechanism that allows the user to turn the data collection on and off?

Yes, the [Firefox telemetry opt-out](https://support.mozilla.org/en-US/kb/share-data-mozilla-help-improve-firefox).

3) If the request is for permanent data collection, is there someone who will monitor the data over time?

Travis and a host of others!

4) Using the **[category system of data types](https://wiki.mozilla.org/Firefox/Data_Collection)** on the Mozilla wiki, what collection type of data do the requested measurements fall under?

Cat 1, technical data

5) Is the data collection request for default-on or default-off?

Default-on

6) Does the instrumentation include the addition of **any *new* identifiers** (whether anonymous or otherwise; e.g., username, random IDs, etc.  See the appendix for more details)?

Yes, an enrollment_id.

7) Is the data collection covered by the existing Firefox privacy notice?

Yes.

8) Does there need to be a check-in in the future to determine whether to renew the data?

No.

9) Does the data collection use a third-party collection tool?

No.

**Comment by travis79** [comment]
Sure, here's a quick breakdown of the enrollment_id:
The `enrollment_id` is generated when the enrollment occurs and is stored with the experiment so long as it remains in the cache and is being reported in telemetry. Once the experiment is ended and we are no longer reporting it in telemetry, the enrollment_id is deleted along with the rest of the experiment info.  

This should remain true even when we start recording to Glean at the Rust layer rather than here at the Kotlin layer.

**Comment by travis79** [comment]
Added the "do not land" while this finishes datareview and I add some tests for the event metrics that were added.

**Review by jhugman — COMMENTED** [review]
This is looking fine, except for a question about opt-ins.

**Comment by litkaemily** [comment]
Approved. Thanks!

Emily

File: components/service/nimbus/docs/metrics.md
Comment (travis79) [reply components/service/nimbus/docs/metrics.md:12]: Unfortunately this is an autogenerated file from the glean_parser step. I've [filed a bug](https://bugzilla.mozilla.org/show_bug.cgi?id=1682583) to address this in the appropriate place.

File: components/service/nimbus/src/main/java/mozilla/components/service/nimbus/Nimbus.kt
Comment (travis79) [reply components/service/nimbus/src/main/java/mozilla/components/service/nimbus/Nimbus.kt:216]: I don't think we can get this information from the `enrollmentChangeEvents` because it only shows the 'enrolled' experiment once so as to only trigger the enrollment event once per client per experiment.  The events in the list are only there to indicate a change of state in the experiment and don't include active experiments whose enrollment state has not changed.

Comment (travis79) [reply components/service/nimbus/src/main/java/mozilla/components/service/nimbus/Nimbus.kt:216]: As for recording telemetry?  If we opt-in for testing, they may be using the Glean debug tools to verify the enrollment or events, so I think it's still a good idea to update the annotations.  Glean shouldn't know or care if we are testing, we could further annotate the ping to indicate that the user "opted-in" here also.

File: buildSrc/src/main/java/Dependencies.kt
Comment (rfk) [inline buildSrc/src/main/java/Dependencies.kt:33]: I believe we need a new release of appservices that fixes the data-migration issue that we encountered during a previous attempt to update to v68. @travis79 can you think of any other outstanding nimbus functionality that we should try to get into such a release?

File: components/service/nimbus/metrics.yaml
Comment (rfk) [inline components/service/nimbus/metrics.yaml:30]: nit: let's remember to update this with details of the data review before landing :-)

File: components/service/nimbus/metrics.yaml
Comment (rfk) [inline components/service/nimbus/metrics.yaml:73]: I recall some discussion about adding a `reason` field to these events, and the underlying event object returned by Rust does have a `reason` field with values `targeting`, `optout` and `error`. Worth adding here, or maybe defer to a followup?

File: components/service/nimbus/src/main/java/mozilla/components/service/nimbus/Nimbus.kt
Comment (rfk) [reply components/service/nimbus/src/main/java/mozilla/components/service/nimbus/Nimbus.kt:216]: On the other hand, it may be useful to do this to test that the expected telemetry is being emitted. We'll also end up calling `Glean.setExperimentActive` on the opted-in experiment on next startup anyway, so I'm not sure it would have much effect if we avoided calling it here.

**Review by rfk — CHANGES_REQUESTED** [review]
This looks good to me!

I'm requesting changes here specifically to help prevent landing with appservices `68.0.0`, which IIRC has a known issue with data migrations. See comments below for more.

File: buildSrc/src/main/java/Dependencies.kt
Comment (rfk) [reply buildSrc/src/main/java/Dependencies.kt:33]: And if there is nothing else urgent pending, would you like to try making a new nimbus-sdk release following the recently-added docs in the README, and I can incorporate it into an appservices release? 😁

Comment (travis79) [reply buildSrc/src/main/java/Dependencies.kt:33]: > I believe we need a new release of appservices that fixes the data-migration issue that we encountered during a previous attempt to update to v68. @travis79 can you think of any other outstanding nimbus functionality that we should try to get into such a release?

Right, this PR was what started all of that and I never updated the dependency. I don't know of anything that we _need_ to have in the update, but the on_telemetry_disabled one would be nice to have.

File: components/service/nimbus/metrics.yaml
Comment (travis79) [reply components/service/nimbus/metrics.yaml:73]: Let's leave it for a followup. I [filed an issue](https://jira.mozilla.com/browse/SDK-155) to track it.

File: buildSrc/src/main/java/Dependencies.kt
Comment (rfk) [reply buildSrc/src/main/java/Dependencies.kt:33]: Thanks! Version `v68.1.0` of appservices has been cut and is currently building, you can check whether it's ready by seeing if the little status dot on [this commit](https://github.com/mozilla/application-services/commit/1d57d7e5703ecaf09132b8161aaffe190eb2e1e6) has changed from yellow to green.

**Review by rfk — APPROVED** [review]
With the new appservices release soon to be available, I'm 👍 on this PR

**Comment by travis79** [comment]
Pinging @jonalmeida for an A-C peer review please! :)

**Review by brizental — APPROVED** [review]

**Comment by travis79** [comment]
Retrying CI, weird failure seems unrelated to this PR (but could be, hence the re-run)

**Comment by travis79** [comment]
@csadilek pointed out that let's not update appservices before the holidays.  Adding the do-not-land until Jan.

**Comment by travis79** [comment]
@csadilek now that we are back from the break, are you okay with landing this PR with the update to appservices?

**Comment by csadilek** [comment]
> @csadilek now that we are back from the break, are you okay with landing this PR with the update to appservices?

Sure, we do want to run a quick smoke test of Fenix though as part of reviews upgrading A-S. That's 10 mins well spent :)

**Comment by csadilek** [comment]
There is (or was) also a way to trigger automated integration tests //cc @grigoryk , but some manual tests are much appreciated.

**Comment by grigoryk** [comment]
bors try

**Comment by grigoryk** [comment]
The `try` push runs `test-ui-browser` which performs a very basic end-to-end on device integration "smoke test" (launches a sample browser with a-s libs configured to be used for storage, at least).

**Comment by grigoryk** [comment]
While the try push failed, `test-ui-browser` did pass. Good! That means we're not crashing at least when the new a-s version is integrated against. Let's do a round of manual testing to be sure.

**Review by jonalmeida — APPROVED** [review]
I did a manual smoke test in Fenix as well as a code review and it looks good to me. 🙂

**Comment by travis79** [comment]
I just rebased this and hopefully it passes this time.

**Comment by travis79** [comment]
@csadilek @grigoryk Aside from the codecov, this looks green.  Do I need to try and satisfy the test coverage now or what do I need to do to proceed with this?  Thanks!

**Comment by jonalmeida** [comment]
Since this PR contains an app services version bump, we should make note of the current state in https://github.com/mozilla-mobile/fenix/issues/17585 before it lands.

**Comment by travis79** [comment]
> Since this PR contains an app services version bump, we should make note of the current state in [mozilla-mobile/fenix#17585](https://github.com/mozilla-mobile/fenix/issues/17585) before it lands.

What does this require?  I would like to get this landed soon as it will be blocking work that we have planned.  I have cycles to work on this and I want to resolve this if possible to avoid it rotting.

**Comment by jonalmeida** [comment]
> .

There is nothing blocking it at the moment, we only need newer a-s release so that we can check in some a-c security patches. Either this PR, or a follow-up can upgrade to the latest a-s version.

**Comment by travis79** [comment]
> > .
> 
> There is nothing blocking it at the moment, we only need newer a-s release so that we can check in some a-c security patches. Either this PR, or a follow-up can upgrade to the latest a-s version.

Awesome @jhugman has a PR which requires a version of A-S that was just released (so newest A-S), maybe it would be better to land that one first and then I can rebase this on his changes.

**Comment by travis79** [comment]
@jhugman Would you care to re-review this now that it is rebased on your PR and has the added recording of the exposure event?  I left the latest changes that add the exposure event as a separate commit to hopefully make it easier to review.  Thanks!

**Pull request merged**
